src/platform/backends/libvirt/libvirt_virtual_machine.cpp

```diff
@@ -378,10 +398,10 @@ std::string mp::LibVirtVirtualMachine::ipv6()
 
 void mp::LibVirtVirtualMachine::wait_until_ssh_up(std::chrono::milliseconds timeout)
 {
-    mp::utils::wait_until_ssh_up(this, timeout);
+    mp::utils::wait_until_ssh_up(this, timeout, std::bind(&LibVirtVirtualMachine::ensure_vm_is_running, this));
 }
 
 void mp::LibVirtVirtualMachine::update_state()
 {
-    monitor->persist_state_for(vm_name);
+    monitor->persist_state_for(vm_name, state);
```

Comment:
I decided to do this because previously, `persist_state_for()` would turn right around and ask the vm for it's state, which really seems unnecessary given the vm is the one requesting this.  And also, having the vm tell the monitor (daemon) the state avoided a pretty serious concurrency issue I was running in to.

Also, `update_state()` is only called when new state is set, so state will always be outdated.